### ISSUE TITLE
aws-node failing readness probe on "fast-paced" nodes

### ISSUE THREAD

**Issue by gabrielbiasi** [aws/amazon-vpc-cni-k8s#1808]
**What happened**:
My aws-node pods starts failing their readness probe on nodes with a **lot of pods been scheduling and terminating**, but NOT on quiet nodes (where my main workloads actually are).

To give you guys some context:

I have a managed node group using spot instances in order to use them as runners for GitHub Actions (ref to project [here](https://github.com/actions-runner-controller/actions-runner-controller)).
These pods act like jobs, after each workflow executed their status goes to `Succeeded` and `Terminating`, in order to be rescheduled as a new runner again.

However, every time the nodes start to "working hard" (2~3 pods been terminating/scheduling per minute) the aws-node pods starts failing the readness probe and all the remaing pods on the given node get stucked on `Pending` or `Terminating`.

I executed `sudo /opt/cni/bin/aws-cni-support.sh` as recommended in other issues, so I got this:

Checking the `plugin.log`, it started to receive connection refused:
```
{"level":"info","ts":"2022-01-12T13:33:21.304Z","caller":"routed-eni-cni-plugin/cni.go:111","msg":"Received CNI add request: ContainerID(520caafcf801703d5ef382c5a1d98b860d73207d36edb832187ad8d23028ef9f) Netns(/var/run/netns/cni-6ebbd319-18a6-a7cc-3f0d-d30e07533e3f) IfName(eth0) Args(IgnoreUnknown=1;K8S_POD_NAMESPACE=runners-3;K8S_POD_NAME=self-hosted-runner-wlqbw-gsjhc;K8S_POD_INFRA_CONTAINER_ID=520caafcf801703d5ef382c5a1d98b860d73207d36edb832187ad8d23028ef9f) Path(/opt/cni/bin) argsStdinData({\"cniVersion\":\"0.3.1\",\"mtu\":\"9001\",\"name\":\"aws-cni\",\"pluginLogFile\":\"/var/log/aws-routed-eni/plugin.log\",\"pluginLogLevel\":\"DEBUG\",\"type\":\"aws-cni\",\"vethPrefix\":\"eni\"})"}
{"level":"debug","ts":"2022-01-12T13:33:21.304Z","caller":"routed-eni-cni-plugin/cni.go:111","msg":"MTU value set is 9001:"}
{"level":"info","ts":"2022-01-12T13:33:21.307Z","caller":"routed-eni-cni-plugin/cni.go:111","msg":"Received add network response for container 520caafcf801703d5ef382c5a1d98b860d73207d36edb832187ad8d23028ef9f interface eth0: Success:true IPv4Addr:\"10.20.25.239\" VPCv4CIDRs:\"10.20.0.0/16\""}
{"level":"debug","ts":"2022-01-12T13:33:21.307Z","caller":"routed-eni-cni-plugin/cni.go:205","msg":"SetupNS: hostVethName=enicde69f1689b, contVethName=eth0, netnsPath=/var/run/netns/cni-6ebbd319-18a6-a7cc-3f0d-d30e07533e3f, deviceNumber=0, mtu=9001"}
{"level":"debug","ts":"2022-01-12T13:33:21.307Z","caller":"driver/driver.go:280","msg":"v4addr: 10.20.25.239/32; v6Addr: <nil>\n"}
{"level":"debug","ts":"2022-01-12T13:33:21.348Z","caller":"driver/driver.go:280","msg":"setupVeth network: disabled IPv6 RA and ICMP redirects on enicde69f1689b"}
{"level":"debug","ts":"2022-01-12T13:33:21.348Z","caller":"driver/driver.go:274","msg":"Setup host route outgoing hostVeth, LinkIndex 75"}
{"level":"debug","ts":"2022-01-12T13:33:21.348Z","caller":"driver/driver.go:274","msg":"Successfully set host route to be 10.20.25.239/0"}
{"level":"info","ts":"2022-01-12T13:33:21.348Z","caller":"driver/driver.go:274","msg":"Added toContainer rule for 10.20.25.239/32"}
{"level":"info","ts":"2022-01-12T13:33:43.786Z","caller":"routed-eni-cni-plugin/cni.go:266","msg":"Received CNI del request: ContainerID(2f5c7921ca9eccb6c404834a9922e813ac7bb451645d3e5ae1d89f16c0e9112f) Netns() IfName(eth0) Args(IgnoreUnknown=1;K8S_POD_NAMESPACE=runners-3;K8S_POD_NAME=self-hosted-runner-wlqbw-4nbph;K8S_POD_INFRA_CONTAINER_ID=2f5c7921ca9eccb6c404834a9922e813ac7bb451645d3e5ae1d89f16c0e9112f) Path(/opt/cni/bin) argsStdinData({\"cniVersion\":\"0.3.1\",\"mtu\":\"9001\",\"name\":\"aws-cni\",\"pluginLogFile\":\"/var/log/aws-routed-eni/plugin.log\",\"pluginLogLevel\":\"DEBUG\",\"type\":\"aws-cni\",\"vethPrefix\":\"eni\"})"}
{"level":"info","ts":"2022-01-12T13:33:43.789Z","caller":"routed-eni-cni-plugin/cni.go:266","msg":"Container 2f5c7921ca9eccb6c404834a9922e813ac7bb451645d3e5ae1d89f16c0e9112f not found"}

[... stop working from here \/ ...]

{"level":"info","ts":"2022-01-12T14:07:19.644Z","caller":"routed-eni-cni-plugin/cni.go:266","msg":"Received CNI del request: ContainerID(4ed621172d692e0e1210386687c863eaf3476a3b65577d3dd1a322493469e415) Netns(/var/run/netns/cni-fae7b698-599b-cb08-edee-6aba72aaaaf9) IfName(eth0) Args(K8S_POD_INFRA_CONTAINER_ID=4ed621172d692e0e1210386687c863eaf3476a3b65577d3dd1a322493469e415;IgnoreUnknown=1;K8S_POD_NAMESPACE=runners-1;K8S_POD_NAME=self-hosted-runner-t9lj2-xlsp2) Path(/opt/cni/bin) argsStdinData({\"cniVersion\":\"0.3.1\",\"mtu\":\"9001\",\"name\":\"aws-cni\",\"pluginLogFile\":\"/var/log/aws-routed-eni/plugin.log\",\"pluginLogLevel\":\"DEBUG\",\"type\":\"aws-cni\",\"vethPrefix\":\"eni\"})"}
{"level":"info","ts":"2022-01-12T14:07:19.645Z","caller":"routed-eni-cni-plugin/cni.go:266","msg":"Received CNI del request: ContainerID(4ed621172d692e0e1210386687c863eaf3476a3b65577d3dd1a322493469e415) Netns(/var/run/netns/cni-fae7b698-599b-cb08-edee-6aba72aaaaf9) IfName(eth0) Args(K8S_POD_NAMESPACE=runners-1;K8S_POD_NAME=self-hosted-runner-t9lj2-xlsp2;K8S_POD_INFRA_CONTAINER_ID=4ed621172d692e0e1210386687c863eaf3476a3b65577d3dd1a322493469e415;IgnoreUnknown=1) Path(/opt/cni/bin) argsStdinData({\"cniVersion\":\"0.3.1\",\"mtu\":\"9001\",\"name\":\"aws-cni\",\"pluginLogFile\":\"/var/log/aws-routed-eni/plugin.log\",\"pluginLogLevel\":\"DEBUG\",\"type\":\"aws-cni\",\"vethPrefix\":\"eni\"})"}
{"level":"error","ts":"2022-01-12T14:07:19.645Z","caller":"routed-eni-cni-plugin/cni.go:266","msg":"Error received from DelNetwork gRPC call for container 4ed621172d692e0e1210386687c863eaf3476a3b65577d3dd1a322493469e415: rpc error: code = Unavailable desc = connection error: desc = \"transport: Error while dialing dial tcp 127.0.0.1:50051: connect: connection refused\""}
{"level":"error","ts":"2022-01-12T14:07:19.645Z","caller":"routed-eni-cni-plugin/cni.go:266","msg":"Error received from DelNetwork gRPC call for container 4ed621172d692e0e1210386687c863eaf3476a3b65577d3dd1a322493469e415: rpc error: code = Unavailable desc = connection error: desc = \"transport: Error while dialing dial tcp 127.0.0.1:50051: connect: connection refused\""}
{"level":"info","ts":"2022-01-12T14:07:21.109Z","caller":"routed-eni-cni-plugin/cni.go:266","msg":"Received CNI del request: ContainerID(4ed621172d692e0e1210386687c863eaf3476a3b65577d3dd1a322493469e415) Netns(/var/run/netns/cni-fae7b698-599b-cb08-edee-6aba72aaaaf9) IfName(eth0) Args(K8S_POD_NAME=self-hosted-runner-t9lj2-xlsp2;K8S_POD_INFRA_CONTAINER_ID=4ed621172d692e0e1210386687c863eaf3476a3b65577d3dd1a322493469e415;IgnoreUnknown=1;K8S_POD_NAMESPACE=runners-1) Path(/opt/cni/bin) argsStdinData({\"cniVersion\":\"0.3.1\",\"mtu\":\"9001\",\"name\":\"aws-cni\",\"pluginLogFile\":\"/var/log/aws-routed-eni/plugin.log\",\"pluginLogLevel\":\"DEBUG\",\"type\":\"aws-cni\",\"vethPrefix\":\"eni\"})"}
{"level":"error","ts":"2022-01-12T14:07:21.109Z","caller":"routed-eni-cni-plugin/cni.go:266","msg":"Error received from DelNetwork gRPC call for container 4ed621172d692e0e1210386687c863eaf3476a3b65577d3dd1a322493469e415: rpc error: code = Unavailable desc = connection error: desc = \"transport: Error while dialing dial tcp 127.0.0.1:50051: connect: connection refused\""}
{"level":"info","ts":"2022-01-12T14:07:37.509Z","caller":"routed-eni-cni-plugin/cni.go:111","msg":"Received CNI add request: ContainerID(17032681e44a340b8b75e0ec6ec8326162e914ba62e0677a208aff0204b3a549) Netns(/var/run/netns/cni-90239816-2687-c527-af81-456a99ce475d) IfName(eth0) Args(IgnoreUnknown=1;K8S_POD_NAMESPACE=runners-3;K8S_POD_NAME=self-hosted-runner-wlqbw-kwt7c;K8S_POD_INFRA_CONTAINER_ID=17032681e44a340b8b75e0ec6ec8326162e914ba62e0677a208aff0204b3a549) Path(/opt/cni/bin) argsStdinData({\"cniVersion\":\"0.3.1\",\"mtu\":\"9001\",\"name\":\"aws-cni\",\"pluginLogFile\":\"/var/log/aws-routed-eni/plugin.log\",\"pluginLogLevel\":\"DEBUG\",\"type\":\"aws-cni\",\"vethPrefix\":\"eni\"})"}
{"level":"debug","ts":"2022-01-12T14:07:37.509Z","caller":"routed-eni-cni-plugin/cni.go:111","msg":"MTU value set is 9001:"}
```

Checking the `ipamd.log` on the same timestamp, no error is presented, but notice that `aws-k8s-agent` gets restarted after a while:
```
{"level":"info","ts":"2022-01-12T11:17:29.375Z","caller":"aws-k8s-agent/main.go:27","msg":"Starting L-IPAMD   ..."}
{"level":"info","ts":"2022-01-12T11:17:29.377Z","caller":"aws-k8s-agent/main.go:42","msg":"Testing communication with server"}
{"level":"info","ts":"2022-01-12T11:17:29.383Z","caller":"aws-k8s-agent/main.go:42","msg":"Successful communication with the Cluster! Cluster Version is: v1.20+. git version: v1.20.7-eks-d88609. git tree state: clean. commit: d886092805d5cc3a47ed5cf0c43de38ce442dfcb. platform: linux/amd64"}

[... after sometime working ...]

{"level":"debug","ts":"2022-01-12T13:44:06.524Z","caller":"ipamd/ipamd.go:646","msg":"Reconcile existing ENI eni-024ca0d88c0ff6037 IP prefixes"}
{"level":"debug","ts":"2022-01-12T13:44:06.524Z","caller":"ipamd/ipamd.go:1347","msg":"Found prefix pool count 0 for eni eni-024ca0d88c0ff6037\n"}
{"level":"debug","ts":"2022-01-12T13:44:06.524Z","caller":"ipamd/ipamd.go:646","msg":"Successfully Reconciled ENI/IP pool"}
{"level":"debug","ts":"2022-01-12T13:44:06.524Z","caller":"ipamd/ipamd.go:646","msg":"IP/Prefix Address Pool stats: total: 28, assigned: 3, cooldownIPs: 0, total prefixes: 0"}

[... notice the huge timestamp gap between these consecutive logs here ...]

{"level":"info","ts":"2022-01-12T14:10:28.760Z","caller":"aws-k8s-agent/main.go:27","msg":"Starting L-IPAMD   ..."}
{"level":"info","ts":"2022-01-12T14:10:28.765Z","caller":"aws-k8s-agent/main.go:42","msg":"Testing communication with server"}
{"level":"info","ts":"2022-01-12T14:10:28.780Z","caller":"aws-k8s-agent/main.go:42","msg":"Successful communication with the Cluster! Cluster Version is: v1.20+. git version: v1.20.7-eks-d88609. git tree state: clean. commit: d886092805d5cc3a47ed5cf0c43de38ce442dfcb. platform: linux/amd64"}
```

I can share the entire output from `aws-cni-support.sh`, if needed.
I already tried to increase the readness timeout, but with no success. 

**Environment**:
- Kubernetes version: EKS version 1.20
- CNI Version: 1.10.1
- OS: Amazon Linux 2
- Kernel: 5.4.156-83.273.amzn2.x86_64

**Comment by Shreya027** [comment]
Hi @gabrielbiasi, can you share the entire output from `aws-cni-support.sh` to `k8s-awscni-triage@amazon.com`. Thanks!

**Comment by jayanthvn** [comment]
@gabrielbiasi - Can you please email the logs bundle to `k8s-awscni-triage@amazon.com`. Also did you try increasing the liveness and readiness probe timeoutSeconds? - https://github.com/aws/amazon-vpc-cni-k8s/blob/master/config/master/aws-k8s-cni.yaml#L146-L163

**Comment by gabrielbiasi** [comment]
Hi @jayanthvn!
Unfortunately, I already tried to increase both liveness and readness probes, but nothing changes.
I just email you guys with the logs bundle of our nodes.

**Comment by Shreya027** [comment]
Thanks @gabrielbiasi , we are looking into the logs, will get back with the update soon.

**Comment by gabrielbiasi** [comment]
@Shreya027 I tried up to 60 seconds and got the same results.

**Comment by gabrielbiasi** [comment]
Hi @Shreya027 and @jayanthvn! Thanks again for taking your time to looking into this.
I noticed that you are preparing the new patch version. I wonder if this patch has anything related to fix this issue?

**Comment by sidewinder12s** [comment]
Any details about what is going on here? We think we might have run into this as well in similar circumstances.

**Comment by jayanthvn** [comment]
Looks like aws-node is starved of CPU and hence on nodes with high pod churn probes are failing. If you expect higher cpu load on the system, CPU resource req and timeout values needs to be adjusted accordingly. Do you have default 10m as cpu requests - https://github.com/aws/amazon-vpc-cni-k8s/blob/master/config/master/aws-k8s-cni.yaml#L213? Did you try increasing it? Also can you please track CPU utilization when there is high pod churn(i.e, when aws-node restarts)?

**Comment by gabrielbiasi** [comment]
Thanks for the insight @jayanthvn . I'm going to perform some tests increasing their resources as well as switching QoS of the pods to Guaranteed, maybe this will stabilize them.

**Comment by jayanthvn** [comment]
Thanks @gabrielbiasi. Please do share the results.

**Comment by gabrielbiasi** [comment]
Hi @jayanthvn  and @Shreya027 !
First, I'd like to thank you for the time you've taken to look at this issue.
Fortunately, today we finished our tests and discovered the root cause, and I would like to share our findings.

Although CPU and RAM usage were high, they didn't reach the saturation point of EC2 instances. However, one very important metric went unnoticed: `burst balance` of the EC2 volumes.

We used `gp2` volumes, and these volumes are subject to consume the burst balance when performing I/O operations. We found that the cluster nodes where the pods were failing had no burst balance remaining and the I/O saturation was at 100%.

By following this [documentation](https://aws.amazon.com/premiumsupport/knowledge-center/ec2-instance-slow-cpu-not-high/), we saw that the solution was to increase the disk sizes, so we have a bigger burst balance, or change the volume type to `io1` and have provisioned IOPS. We chose the second option, and the pods stopped failing their readness and liveness probes.

I hope this solution will help people in the future to investigate not only CPU and RAM, but also volume and network metrics.
